src/main/scala/mesosphere/marathon/api/TaskKiller.scala

```diff
@@ -100,8 +103,15 @@ class TaskKiller @Inject() (
       toKill = appTasks
     )
 
-    appTasks.keys.find(id => !instanceTracker.hasSpecInstancesSync(id))
-      .map(id => Future.failed(PathNotFoundException(id)))
-      .getOrElse(killTasks)
+    val idHasSpec = Future.sequence(appTasks.keys.map(id => async {
+      id -> await(instanceTracker.hasSpecInstances(id))
+    }))
+
+    async {
+      val hasSpec = await(idHasSpec).groupBy(_._2)
```

Comment:
I am not a big fan of tuple property syntax (`_2`).
If you use partition like suggested above, you don't need this.